Store claimants names as separate fields

robbiepaul opened the pull request:
HMRC expect employees names to be reported as three separate fields: Forename, Second forename, and Surname. Given this, and given that we receive three separate names from Verify, we should probably store claimants names as three separate fields. 

Added a data migration using the gem [`data_migrate`](https://github.com/ilyakatz/data-migrate) - it seems well supported with active development. The features it provides will allow us to confidently migrate and modify data.

There was one annoyance I found with it, it doesn't create a db schema migration for the `data_migrations` table it generates, so I created one manually. There is an open issue on the gem repo about it: https://github.com/ilyakatz/data-migrate/issues/105

pezholio commented on app/views/claim_mailer/submitted.text.erb:
```diff
@@ -1,4 +1,4 @@
-Dear <%= @claim.full_name %>,
+Dear <%= @claim.first_name %>,
```
Do we want to address them as their full name here? It might be good to have a `full_name` method that pieces the name parts together

pezholio commented on spec/presenters/tslr_claims_csv_spec.rb:
```diff
@@ -55,7 +57,9 @@
         eligibility.current_school_name,
-        claim.full_name,
+        claim.first_name,
+        nil,
```
Might be a good idea to check the middle name is in the CSV, even though it's optional. It's bitten me in the past

pezholio commented on db/migrate/20190829145138_create_data_migrations.rb:
```diff
@@ -0,0 +1,5 @@
+class CreateDataMigrations < ActiveRecord::Migration[5.2]
+  def change
+    create_table :data_migrations, primary_key: "version", id: :string unless table_exists? :data_migrations
```
I know it's not A Thing in migrations usually, but can we wrap the arguments to `create_table` in brackets, given that we have an `unless` call that the end. Or better still, have a `return if table_exists? :data_migrations` guard clause on the first line

pezholio commented on README.md:
```diff
@@ -118,6 +118,17 @@ Code linting is performed using:
+The migrations are stored in the `db/data` folder.
+
+When the service deploys it will run these data migrations at the same time as
+the schema ones.
```
Can you add instructions on how to generate a migration, and how to run the data migrations at the same time as the DB migrations? (i.e. run `rake db:migrate:with_data`). I know it's in the gem docs, but makes sense to be here too

robbiepaul replied:
Yeah sure, my worry was that the commands may get out of date as the gem updates but we can cross that bridge when it comes to it 👍

pezholio commented on db/data/20190829135327_backfill_first_middle_surnames.rb:
```diff
@@ -0,0 +1,26 @@
+class BackfillFirstMiddleSurnames < ActiveRecord::Migration[5.2]
+  def up
+    claims_with_full_names.each do |claim|
+      parser = Claim::VerifyResponseParametersParser.new(claim.verify_response)
```
Nice, having the parser broken out is super useful here 😃

robbiepaul replied:
Yeah it made things super simple 👍 thanks @tekin

robbiepaul commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Store claimants' names as separate fields](https://trello.com/c/0tTLbfZq/614-store-claimants-names-as-separate-fields)

robbiepaul commented:
Ready for another look

tekin commented on bin/docker-entrypoint:
```diff
@@ -11,8 +11,8 @@ else
-    echo "Running the migrations"
-    rails db:migrate
+    echo "Running the schema & data migrations"
+    rails db:migrate:with_data
```
Looking at the definition for this rake task makes me wary. It looks like it re-implements `db:migrate`. Unless there's good reason, I'd feel more comfortable running `db:migrate data:migrate` separately.

robbiepaul replied:
👍

tekin commented on db/migrate/20190829094432_add_first_middle_last_names_to_claim.rb:
```diff
@@ -0,0 +1,7 @@
+class AddFirstMiddleLastNamesToClaim < ActiveRecord::Migration[5.2]
+  def change
+    add_column :claims, :first_name, :string, limit: 100
```
How were these limits chosen?

robbiepaul replied:
I just halved the full name one, happy to change it though

tekin replied:
no, It should be fine.

tekin commented on README.md:
```diff
@@ -118,6 +118,22 @@ Code linting is performed using:
+  `rails db:migrate:with_data`
+
+When the service deploys it will run these data migrations at the same time as
+the schema ones.
```
can we reword this – if we run `rails db:migrate` followed by `rails data:migrate` sequentially rather than using `rails db:migrate:with_data`, which looks like it intermingles the two. I'd rather we had a workflow that required schema migrations were run first and then data migrations run after. that does some changes will require splitting across more than one release (i.e. a follow up PR to drop the now-unused column), but I think this approach is safer and less likely to result in irreversible accidental data loss.

robbiepaul replied:
Yeah that makes sense, I've made the change

robbiepaul commented:
@tekin Ready for another review. I added the separation of `db:migrate` and `data:migrate`, updated the readme. 

I also had to add a commit that ignores some warnings from standard for the generated file 55e824cb1754a7c9f5c10a68d934e56473ea815f
